### PR TITLE
feat: openai_api_compatible support config stream_mode_delimiter

### LINKED ISSUE #2144: the openai-compatible model's  answer is blank with streaming mode
### Self Checks

- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to file this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).

### Dify version

 0.4.9

### Cloud or Self Hosted

Self Hosted (Docker)

### Steps to reproduce

I add a openai-compatible model(rwkv docker service), the api is working fine;
but in dify, the answer is not show.

I test blocking, the answer is show.
```bash
curl --location --request POST 'http://my-domain/v1/chat-messages' \
--header 'Authorization: Bearer app-xxx' \
--header 'Content-Type: application/json' \
--data-raw '{
    "inputs": {},
    "query": "who are you",
    "response_mode": "blocking",
    "conversation_id": "",
    "user": "dify-api"
}'
{"event": "message", "task_id": "d0ecb553-fa53-46e7-a76a-83f962795fa0", "id": "79d7e38e-9877-46bd-aad8-b2aa039ceb04", "message_id": "79d7e38e-9877-46bd-aad8-b2aa039ceb04", "mode": "chat", "answer": "\u6211\u662f\u4e00\u540d\u5168\u6808\u5f00\u53d1\u4eba\u5458\uff0c\u64c5\u957fpython,kotlin,markdown,\u719f\u6089flink,spark\u8ba1\u7b97\u7cfb\u7edf\uff0c\u5bf9mysql,elasticsearch\u7b49\u5b58\u50a8\u7cfb\u7edf\u6709\u5b9e\u6218\u7ecf\u9a8c\u3002\n", "metadata": {"usage": {"prompt_tokens": 72, "prompt_unit_price": "0", "prompt_price_unit": "0", "prompt_price": "0.0000000", "completion_tokens": 48, "completion_unit_price": "0", "completion_price_unit": "0", "completion_price": "0.0000000", "total_tokens": 120, "total_price": "0.0000000", "currency": "USD", "latency": 2.3984292540117167}}, "created_at": 1706019858, "conversation_id": "997fb182-88d7-46c9-a4d7-672437a6fd5a"}
```
blocking

I test the streaming, the answer is blank
```bash
curl --location --request POST 'http://my-domain/v1/chat-messages' \
--header 'Authorization: Bearer app-xxx' \
--header 'Content-Type: application/json' \
--data-raw '{
    "inputs": {},
    "query": "who are you",
    "response_mode": "streaming",
    "conversation_id": "",
    "user": "dify-api"
}'
event: ping

data: {"event": "message", "id": "f3f16af0-a327-43e8-9071-d143575f57d1", "task_id": "3e070c26-8d89-446e-bd9e-45200136eaa0", "message_id": "f3f16af0-a327-43e8-9071-d143575f57d1", "answer": "", "created_at": 1706019951, "conversation_id": "35e7467a-dbef-47c4-8336-d4330bc2b82a"}

data: {"event": "message_end", "task_id": "3e070c26-8d89-446e-bd9e-45200136eaa0", "id": "f3f16af0-a327-43e8-9071-d143575f57d1", "message_id": "f3f16af0-a327-43e8-9071-d143575f57d1", "conversation_id": "35e7467a-dbef-47c4-8336-d4330bc2b82a", "metadata": {"usage": {"prompt_tokens": 106, "prompt_unit_price": "0", "prompt_price_unit": "0", "prompt_price": "0.0000000", "completion_tokens": 0, "completion_unit_price": "0", "completion_price_unit": "0", "completion_price": "0.0000000", "total_tokens": 106, "total_price": "0.0000000", "currency": "USD", "latency": 14.101711608003825}}}
```
blank-streaming

### ✔️ Expected Behavior

the streaming mode should working fine to show the answer

### ❌ Actual Behavior

with streaming mode   the answer is blank.
also the preview page the answer  is blank too

### PR DESCRIPTION
**Pull request by geosmart**
ref https://github.com/langgenius/dify/issues/2144

### CODE REVIEW

**Review by crazywoola — DISMISSED** [review]

**Comment by crazywoola** [comment]
Please add my wechat: crazyphage. I will add you to the contributors group.

File: api/core/model_runtime/model_providers/openai_api_compatible/llm/llm.py [hunk 1]
```diff
@@ -343,22 +343,31 @@ def create_final_llm_result_chunk(index: int, message: AssistantPromptMessage, f
                 )
             )
 
-        for chunk in response.iter_lines(decode_unicode=True, delimiter='\n\n'):
+        # delimiter for stream response, need unicode_escape
+        import codecs
+        delimiter = credentials["stream_mode_delimiter"]
+        delimiter = codecs.decode(delimiter, "unicode_escape")
```
Comment (geosmart) [inline api/core/model_runtime/model_providers/openai_api_compatible/llm/llm.py:349]: > the encrypted_config stored in db encoded the stream_mode_delimiter,:\\\r\\\n\\\r\\\n
{
    "api_key": "xxx",
    "endpoint_url": "http://192.168.68.100:18000/v1/",
    "mode": "chat",
    "context_size": "2000",
    "max_tokens_to_sample": "2100",
    "stream_mode_delimiter": "\\r\\n\\r\\n",
    "stream_mode_exclude_chunks": "[DONE]"
}

so need to decode

File: api/tests/integration_tests/model_runtime/openai_api_compatible/test_llm.py [hunk 2]
```diff
@@ -73,7 +73,9 @@ def test_invoke_stream_model():
         credentials={
             'api_key': os.environ.get('TOGETHER_API_KEY'),
             'endpoint_url': 'https://api.together.xyz/v1/',
-            'mode': 'chat'
+            'mode': 'chat',
+            "stream_mode_delimiter": "\\n\\n",
+            "stream_mode_exclude_chunks": "[DONE]"
```
Comment (takatost) [inline api/tests/integration_tests/model_runtime/openai_api_compatible/test_llm.py]: Can you add a test to check if it works without the stream_mode_delimiter and stream_mode_exclude_chunks?

**Review by takatost — DISMISSED** [review]

**Comment by bowenliang123** [comment]
Great feature.
It would be perfect if we could have a test button to verify whether the delimiter works well.

File: api/tests/integration_tests/model_runtime/openai_api_compatible/test_llm.py (hunk 2 above)
Comment (geosmart) [reply api/tests/integration_tests/model_runtime/openai_api_compatible/test_llm.py]: done  ,get the config with default

delimiter = credentials.get("stream_mode_delimiter", "\r\n")

File: api/core/model_runtime/model_providers/openai_api_compatible/llm/llm.py [hunk 3]
```diff
@@ -343,32 +343,44 @@ def create_final_llm_result_chunk(index: int, message: AssistantPromptMessage, f
...
                     continue
 
                 choice = chunk_json['choices'][0]
+                finish_reason = chunk_json['choices'][0].get('finish_reason')
                 chunk_index += 1
 
                 if 'delta' in choice:
                     delta = choice['delta']
                     if delta.get('content') is None or delta.get('content') == '':
-                        continue
```
Comment (geosmart) [inline api/core/model_runtime/model_providers/openai_api_compatible/llm/llm.py:371]: rwkv model streaming response sample:

when the content is None, the finish reason maybe is not None
```bash
data: {"object": "chat.completion.chunk", "model": "RWKV-5-World-1B5-v2-20231025-ctx4096", "choices": [{"delta": {"content": " to"}, "index": 0, "finish_reason": null}]}

data: {"object": "chat.completion.chunk", "model": "RWKV-5-World-1B5-v2-20231025-ctx4096", "choices": [{"delta": {"content": "!"}, "index": 0, "finish_reason": null}]}

data: {"object": "chat.completion.chunk", "model": "RWKV-5-World-1B5-v2-20231025-ctx4096", "choices": [{"delta": {}, "index": 0, "finish_reason": "stop"}]}

data: [DONE]
```

**Comment by guchenhe** [comment]
lgtm, good work!

**Review by guchenhe — APPROVED** [review]

**Pull request merged**